### PR TITLE
littlefs : Upgrade to v2.4.1 in west manifest

### PR DESCRIPTION
**Pull request by poojamg**
1. Upgrade west sha id to point to https://github.com/zephyrproject-rtos/littlefs/pull/7 ( upgrade littleFS to v2.4.1)

### CODE REVIEW

File: west.yml
Comment (dleach02) [inline west.yml:217]: To reference a PR in a module the syntax here would be:

`revision: pull/7/head
`

File: west.yml
Comment (dleach02) [inline west.yml:223]: I'm not sure what the rest of this is.

File: west.yml
Comment (dleach02) [inline west.yml:217]: Sorry I didn't notice this before but you need to modify the existing entry for littlefs in this manifest. Only need to change the "revision" line on that entry to point to your PR in that module. What you did here was add another entry for littlefs so now CI is breaking on this

**Comment by Laczen** [comment]
@poojamg, how are the options (like LFS_READONLY, LFS_THREADSAFE, ...) defined?

File: subsys/fs/littlefs_fs.c
Comment (de-nordic) [inline subsys/fs/littlefs_fs.c:363]: This `#ifndef LFS_READONLY` is within block that is already included via `#ifndef LFS_READONLY`.

**Comment by de-nordic** [comment]
> @poojamg, how are the options (like LFS_READONLY, LFS_THREADSAFE, ...) defined?

It does not seem to be defined anywhere.

**Review by de-nordic — CHANGES_REQUESTED** [review]
These should be two commits: one for the west update and one for the LFS_READONLY flag.
Additionally the read-only implementation is incomplete so please re-consider doing two separate PRs for updating the LittleFS and adding new features.

**Comment by Laczen** [comment]
@poojamg, @de-nordic littlefs seems to have a way to override the lfs_util.h using a compilation option `-DLFS_CONFIG=lfs_config.h` that provides a method to override certain settings and to enable options. It would be good if these options would be driven by zephyr `CONFIG_...` options and not to pull them in by modifying lfs_util.h in the module folder. This will require more work and is something that should be in a separate PR.

@poojamg, so as @de-nordic proposed it would indeed be better to first update littlefs in this PR without any extra options and afterwards look for the best way to enable these options by creating a zephyr lfs_config.h (using Kconfig options) and overriding littlefs's default setup.

**Comment by de-nordic** [comment]
> @poojamg, @de-nordic littlefs seems to have a way to override the lfs_util.h using a compilation option `-DLFS_CONFIG=lfs_config.h` that provides a method to override certain settings and to enable options. It would be good if these options would be driven by zephyr `CONFIG_...` options and not to pull them in by modifying lfs_util.h in the module folder. This will require more work and is something that should be in a separate PR.

This can be, probably, resolved with changes to the CMakeLists.txt for the fs subsystem.

**Comment by Laczen** [comment]
@poojamg, @de-nordic, it seems that modules/fs/littlefs is using a (zephyr) modified `lfs_util.h`, does this influence the importing of the module ?

**Comment by de-nordic** [comment]
> @poojamg, @de-nordic, it seems that modules/fs/littlefs is using a (zephyr) modified `lfs_util.h`, does this influence the importing of the module ?

Not exactly. But since you mention it @Laczen I think that before we even consider the update, we should evaluate possibility of using "vanilla" little fs, by following the instruction within lfs_util.h.
In short: create our own lfs_util.h (the lfs_config.h) within subsys/fs/ (or whatever) and use the CMakeLists.txt to add it to compilation (the -D thing) and try it on vanilla LittleFS. If that works we could either revert Zephyr changes from LittleFS fs module or replace it completely with vanilla Littlefs (force push).

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
Looking at `.unlink = littlefs_unlink`
Need to set this to NULL if `LFS_READONLY` is defined.

**Comment by Laczen** [comment]
> > @poojamg, @de-nordic, it seems that modules/fs/littlefs is using a (zephyr) modified `lfs_util.h`, does this influence the importing of the module ?
> 
> Not exactly. But since you mention it @Laczen I think that before we even consider the update, we should evaluate possibility of using "vanilla" little fs, by following the instruction within lfs_util.h. In short: create our own lfs_util.h (the lfs_config.h) within subsys/fs/ (or whatever) and use the CMakeLists.txt to add it to compilation (the -D thing) and try it on vanilla LittleFS. If that works we could either revert Zephyr changes from LittleFS fs module or replace it completely with vanilla Littlefs (force push).

@de-nordic, I agree, I was trying to evaluate this by modifying the CMakeLists.txt under`modules/fs/littlefs` as this is needed anyhow. The only needed modification is `add_compile_options(-DLFS_CONFIG=lfs_config_zephyr.h)`. I don't understand what/who generates this CMakeLists.txt, but by adding at the same time the `lfs_config_zephyr.h` should make it possible to use vanilla littlefs.

**Comment by de-nordic** [comment]
> 
> @de-nordic, I agree, I was trying to evaluate this by modifying the CMakeLists.txt under`modules/fs/littlefs` as this is needed anyhow. The only needed modification is `add_compile_options(-DLFS_CONFIG=lfs_config_zephyr.h)`. I don't understand what/who generates this CMakeLists.txt, but by adding at the same time the `lfs_config_zephyr.h` should make it possible to use vanilla littlefs.

OK, so "total vanilla" will not be possible because we will need the CMakeLists.txt (this is Zephyr's and it is done by hand) and zephyr/module.yml, but we can avoid modifying source code which is always a plus.

**Comment by de-nordic** [comment]
We should look into https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/lz4 - this is example of module where CMakeLists.txt is stored within Zephyr code not in module; I have looked at other modules under https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/ and it seems that some of them carry additional code here, so maybe we could move the CMakeListst.txt out of LittleFS module, ant he lfs_config_zephyr.h into the https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/littlefs ?

**Comment by Laczen** [comment]
> We should look into https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/lz4 - this is example of module where CMakeLists.txt is stored within Zephyr code, not in the module; I have looked at other modules under https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/ and it seems that some of them carry additional code here, so maybe we could move the CMakeListst.txt out of LittleFS module, ant he lfs_config_zephyr.h into the https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/littlefs ?

Yes, and we could add the `lfs_config_zephyr.h` together with the `CMakeLists.txt` into the zephyr tree. The only caveat I found is that it is needed to also define `lfs_crc` in the `lfs_config_zephyr.h` file, but this is just a call to crc32_ieee_update().

**Comment by dkalowsk** [comment]
@de-nordic and @Laczen I see earlier in the thread the proposal that this PR be held back until other Zephyr build infrastructure changes can be made.  Any specific reason(s) why?  I get that optimally the two would be the most desirable, but it does seem like one is a functional upgrade and the other is an enhancement upgrade that can be done in parallel.

**Comment by poojamg** [comment]
> @poojamg, @de-nordic littlefs seems to have a way to override the lfs_util.h using a compilation option `-DLFS_CONFIG=lfs_config.h` that provides a method to override certain settings and to enable options. It would be good if these options would be driven by zephyr `CONFIG_...` options and not to pull them in by modifying lfs_util.h in the module folder. This will require more work and is something that should be in a separate PR.
> 
> @poojamg, so as @de-nordic proposed it would indeed be better to first update littlefs in this PR without any extra options and afterwards look for the best way to enable these options by creating a zephyr lfs_config.h (using Kconfig options) and overriding littlefs's default setup.

I agree, today mbedtls is also doing the same refer to  CONFIG_MBEDTLS_CFG_FILE

**Comment by poojamg** [comment]
> https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/lz4

**Comment by Laczen** [comment]
@dkalowsk, @poojamg, for me it is OK not to postpone the upgrade of littlefs. If the PR is limited just to the upgrade (so no use LFS_READONLY flag) we can do the other changes later on.

@poojamg, could you limit the PR just to the upgrade of littlefs to v2.4.1 ?

@de-nordic, I see no reason to postpone the upgrade to littlefs v2.4.1 until "vanilla" littlefs can be used.

**Comment by de-nordic** [comment]
> 
> @de-nordic, I see no reason to postpone the upgrade to littlefs v2.4.1 until "vanilla" littlefs can be used.

The only argument I have is that we can clean up what we have and then bring new. But will not oppose if it will be decided to do otherwise by other reviewers.

**Review by Laczen — APPROVED** [review]

**Comment by Laczen** [comment]
@poojamg, I think the problem you are having is inside littlefs: `lfs_mlist_isopen()` is defined but unused, this results in build errors.

**Comment by poojamg** [comment]
> @poojamg, I think the problem you are having is inside littlefs: `lfs_mlist_isopen()` is defined but unused, this results in build errors.

This is fixed, please can you help approve if PR is good to merge ?

File: tests/subsys/settings/functional/src/settings_basic_test.c
Comment (Laczen) [inline tests/subsys/settings/functional/src/settings_basic_test.c:16]: This should not be done in this PR, is it required ?

**Comment by Laczen** [comment]
@poojamg, it seems that the removal of the log command in `settings_basic_test.c` is required to pass the tests. So please add the removal to your PR.

**Comment by poojamg** [comment]
> @poojamg, it seems that the removal of the log command in `settings_basic_test.c` is required to pass the tests. So please add the removal to your PR.

Yes, it was required, but I reworked the wonky LOG_MODULE_REGISTER / DECLARE, this will still need a good kconfig clean up for later

**Comment by Laczen** [comment]
@de-nordic, @nvlsianpu, could you have another look at the PR.

The removal of the logging lines could be in a separate PR but as these lines are not used anyhow I don't see a problem with removing them here. There seems to be a more general problem with how the logging is integrated in littlefs (trough lfs_util.h which is included by lfs.h which is included by littlefs.h), you cannot have two `LOG_MODULE_REGISTER` lines. Normally this is not a problem as the `LOG_MODULE_REGISTER` is done in a c file but how to achieve this with littlefs ?

**Comment by dkalowsk** [comment]
@de-nordic and @nvlsianpu any input?

**Comment by dleach02** [comment]
@de-nordic and @nvlsianpu, the PR needs your attention to the change requests you have blocking this. After 14 days of no response your blocks will be removed to allow progress

https://docs.zephyrproject.org/latest/development_process/dev_env_and_tools.html#reviewers-should-keep-track-of-pull-requests-they-have-provided-feedback-to

**Comment by de-nordic** [comment]
> @de-nordic and @nvlsianpu, the PR needs your attention to the change requests you have blocking this. After 14 days of no response your blocks will be removed to allow progress
> 
> https://docs.zephyrproject.org/latest/development_process/dev_env_and_tools.html#reviewers-should-keep-track-of-pull-requests-they-have-provided-feedback-to

There is no movement in the zephyrproject-rtos/littlefs#7; this PR is only manifest update that has incorrect sha because the LittleFS update did not get merged yet.

**Comment by dleach02** [comment]
> 
> There is no movement in the [zephyrproject-rtos/littlefs#7](https://github.com/zephyrproject-rtos/littlefs/pull/7); this PR is only manifest update that has incorrect sha because the LittleFS update did not get merged yet.

@de-nordic I understand that but you still have a block/change request on this particular PR. The module PR is likely ready to merge so I'm trying to help this contributor get this one unblocked. So what do you need from this contributor or the project to clear your block on this PR or more specifically, did the changes that @poojamg made after your change request meet your request?

**Comment by dleach02** [comment]
@poojamg, your PR to the littlefs module has been merged. Please update this one to point to the new SHA.

**Comment by poojamg** [comment]
> @poojamg, your PR to the littlefs module has been merged. Please update this one to point to the new SHA.

@dleach02 Thank you! I have updated to the new SHA.

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks!

**Comment by dleach02** [comment]
@nvlsianpu, Need your attention to determine if your change request has been addressed

**Comment by poojamg** [comment]
@dleach02, can we move forward to merge?

**Comment by dleach02** [comment]
The policy is to give reviewers 14 days to respond. I'm waiting on a response from @nvlsianpu before dismissing their review. Tomorrow will be 14 days.

**Comment by nvlsianpu** [comment]
@dleach02 I'm looking into this now.

**Review by nvlsianpu — APPROVED** [review]
I saw that me request was not relevant anymore since this PR was limited to littlefs module upgrade - which is fine.

Thanks for taking that!

**Comment by dleach02** [comment]
thank you @nvlsianpu

**Comment by poojamg** [comment]
Thank you all!

**Pull request merged**
